azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +101,232 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+            expected_type = dict
+        if expected_type == str:
+            expected_type = (str, ustr)
+
+        # Map from python types to the types we want to reference in error messages
+        type_in_err_msg = {
+            dict: "object",
+            (str, ustr): "string",
+            bool: "boolean"
+        }
+
+        # Raise error if attribute is not the expected type
+        if not isinstance(value, expected_type):
+            if parent_attribute_name is None:
+                msg = ("invalid type {0} for attribute '{1}', please change to {2}."
+                       .format(type(value).__name__, key, type_in_err_msg.get(expected_type)))
+            else:
+                msg = ("invalid type {0} for attribute '{1}' in section '{2}', please change to {3}."
+                       .format(type(value).__name__, key, parent_attribute_name,  type_in_err_msg.get(expected_type)))
+
+            raise InvalidPolicyError(msg)
+
+        return value
+
+
+class ExtensionPolicyEngine(_PolicyEngine):
+
+    def should_allow_extension(self, extension_to_check):
+        """
+        Return whether we should allow extension download based on policy.
+        extension_to_check is expected to be an Extension object.
+
+        If policy feature not enabled, return True.
+        If allowListedExtensionsOnly=true, return true only if extension present in "extensions" allowlist.
+        If allowListedExtensions=false, return true always.
+        """
+        if not self.policy_enforcement_enabled:
+            return True
+
+        allow_listed_extension_only = self._policy.get("extensionPolicies").get("allowListedExtensionsOnly")
+        extension_allowlist = self._policy.get("extensionPolicies").get("extensions")
+
+        should_allow = not allow_listed_extension_only or extension_allowlist.get(extension_to_check.name) is not None
+        return should_allow
+
+    def should_enforce_signature_validation(self, extension_to_check):
+        """
+        Return whether we should enforce signature based on policy.
+        extension_to_check is expected to be an Extension object.
+
+        If policy feature not enabled, return False.
+        Individual policy takes precedence over global - if individual signing policy present, return true/false based on
+        individual policy. Else, return true/false based on global policy.
+        """
+        if not self.policy_enforcement_enabled:
+            return False
+
+        global_signature_required = self._policy.get("extensionPolicies").get("signatureRequired")
+        individual_policy = self._policy.get("extensionPolicies").get("extensions").get(extension_to_check.name)
+        if individual_policy is None:
+            return global_signature_required
+        else:
+            return individual_policy.get("signatureRequired")
 
+    # TODO: Consider adding a function should_download_extension() combining should_allow_extension() and
```

Comment:
remove TODO; should_enforce_signature_validation does not determine whether should_download_extension

Reply:
done